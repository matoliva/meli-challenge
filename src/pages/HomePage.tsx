import {products} from '../models/products'
import {ProductList} from '../components/ProductList'

export const HomePage = () => {
  return <ProductList items={products[0].items} />
}
