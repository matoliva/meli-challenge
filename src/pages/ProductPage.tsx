import {useEffect, useState} from 'react'
import {useParams} from 'react-router'
import {Spinner} from '../components/Spinner'
import {useFetchAll} from '../hooks/useFetchAll'
import {IItemMapped} from './HomePage'

interface IItemMappedExt extends IItemMapped {
  sold_quantity: number
  mobilePicture: string
  desktopPicture: string
  plain_text: string
}

export const ProductPage = () => {
  const {id: productId = ''} = useParams()

  const {apiData, isLoading, serverError} = useFetchAll(
    `https://api.mercadolibre.com/items/${productId}`,
    `https://api.mercadolibre.com/items/${productId}/description`,
  )

  const [item, setItem] = useState<IItemMappedExt>({
    id: '',
    title: '',
    price: 0,
    currency: '',
    condition: '',
    address: '',
    sold_quantity: 0,
    picture: '',
    mobilePicture: '',
    desktopPicture: '',
    plain_text: '',
  })

  useEffect(() => {
    const mapDatafromServer = () => {
      let dataMaped: any
      if (apiData) {
        dataMaped = {
          id: apiData.id,
          title: apiData.title,
          price: apiData.price,
          currency: apiData.currency_id,
          address: apiData?.address?.city_name,
          picture: apiData.thumbnail,
          condition: apiData.condition,
          sold_quantity: apiData.sold_quantity,
          mobilePicture: apiData.pictures[0].url,
          desktopPicture: apiData.pictures[2].url,
          plain_text: apiData.plain_text,
        }
        setItem(dataMaped)
      }
    }
    mapDatafromServer()
  }, [apiData])

  const {
    title,
    price,
    currency,
    picture,
    condition,
    sold_quantity,
    mobilePicture,
    desktopPicture,
    plain_text,
  } = item

  const currencyMap = currency === 'ARS' ? '$' : 'uSd'
  //TODO: I have to code a hook that check if it's a desktop or mobile screen
  return !isLoading ? (
    <article className="product-page">
      <div className="product-page__product">
        <img src={desktopPicture} alt={title} />
        <div className="product-page__product--buy">
          <p>{`${condition} - ${sold_quantity} vendidos`}</p>
          <h3>{title}</h3>
          <p className="product-page__product--price">
            {currencyMap} {price}
          </p>
          <button className="btn pointer">Comprar</button>
        </div>
      </div>
      <div className="product-page__description">
        <h2>Descripci√≥n del producto</h2>
        <p>{plain_text}</p>
      </div>
    </article>
  ) : (
    <Spinner />
  )
}
