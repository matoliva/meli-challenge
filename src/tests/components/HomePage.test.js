import {render, waitFor} from '@testing-library/react'
import {BrowserRouter} from 'react-router-dom'
import {HomePage} from '../../pages/HomePage'

describe('Home Page', () => {
  beforeEach(() => {})

  it('renders correctly', () => {
    const {getByTestId} = render(
      <BrowserRouter>
        <HomePage />
      </BrowserRouter>,
    )
    const homePage = getByTestId('home-page')
    expect(homePage).toMatchSnapshot()
  })

  it('should called an async api when the component is rendering', async () => {
    const fakeResponse = {
      id: 1,
      title: 'test title',
      price: 1000,
      currency_id: 'ARS',
      address: {city_name: 'Mar del Plata'},
      thumbnail: 'http://image',
      condition: 'new',
    }

    const mockFetch = Promise.resolve({
      json: () => Promise.resolve(fakeResponse),
    })
    const mockedFetch = jest
      .spyOn(window, 'fetch')
      .mockImplementationOnce(() => mockFetch)
    await waitFor(async () => {
      render(
        <BrowserRouter>
          <HomePage />
        </BrowserRouter>,
      )
    })
    expect(mockedFetch).toHaveBeenCalledTimes(1)
  })
})
