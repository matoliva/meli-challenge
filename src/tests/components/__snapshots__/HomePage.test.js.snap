// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page renders correctly 1`] = `
<div
  data-testid="home-page"
>
  <div
    class="sk-chase"
  >
    <div
      class="sk-chase-dot"
    />
    <div
      class="sk-chase-dot"
    />
    <div
      class="sk-chase-dot"
    />
    <div
      class="sk-chase-dot"
    />
    <div
      class="sk-chase-dot"
    />
    <div
      class="sk-chase-dot"
    />
  </div>
</div>
`;
